pkg/solidity-utils/contracts/math/BasePoolMath.sol

```diff
@@ -291,9 +300,13 @@ library BasePoolMath {
         // Calculate the new invariant ratio by dividing the new invariant by the old invariant.
         // Calculate the new proportional balance by multiplying the new invariant ratio by the current balance.
         // Calculate the taxable amount by subtracting the new balance from the equivalent proportional balance.
-        uint256 currentInvariant = pool.computeInvariant(currentBalances);
+        // We round `currentInvariant` up as it affects the calculated `bptAmountIn` directly (see below).
+        uint256 currentInvariant = pool.computeInvariant(currentBalances, Rounding.ROUND_UP);
         // We round invariant ratio up (see reason below).
-        uint256 invariantRatio = pool.computeInvariant(newBalances).divUp(currentInvariant);
+        // This invariant ratio could be rounded up even more by rounding `currentInvariant` down. But since it only
+        // affects the taxable amount and the fee calculation, whereas `currentInvariant` affects BPT in more directly,
+        // we use `currentInvariant` rounded up here as well.
```

Comment:
Would this (rounding) reduce the difference in fees between "paths" then? So we're making it "fairer" by reducing the delta in this direction, while rounding the invariant itself protects against unfavorable "round trips"

Reply:
More or less.

The most important thing is that the overall amount we ask from the sender in this case makes the roundtrip lossy. Fee calculation is always behind a fee percentage which is < 1, so if you round in detriment of the fees but you still ask more tokens from the user it's not necessarily that you're making the indirect math 'fairer'.

The problem here is that we're computing 2 invariants already to get the invariant ratio. My understanding is that the way in this PR is the one that favors the protocol the most computing only these 2 invariants, which is the goal.

We could round in the protocol's favor even more by favoring the rounding for fees as well, but that would require computing one extra invariant which is too much to ask.